File: ex04/ft_print_comb.c
/*
 * C version for reference
 * Compile: gcc ft_print_comb.c -o ft_print_comb
 * Run: ./ft_print_comb
 */

#include <unistd.h>

void ft_print_comb(void) {
  int first = 0;
  int second = 0;
  int third = 0;
  int first_printed = 0; // Track if we've printed at least one combination

  for (first = 0; first <= 9; first++) {
    for (second = 0; second <= 9; second++) {
      for (third = 0; third <= 9; third++) {
        // Check constraint: digits are different and in ascending order
        if (first < second && second < third) {
          // Print comma and space before each combination except the first
          if (first_printed) {
            write(1, ", ", 2);
          }
          first_printed = 1;

          // Print the three digits
          write(1, &(char){'0' + first}, 1);
          write(1, &(char){'0' + second}, 1);
          write(1, &(char){'0' + third}, 1);
        }
      }
    }
  }
  write(1, "\n", 1);
}

int main(void) {
  ft_print_comb();
  return (0);
}

File: ex04/ft_print_comb.ml
(*
 * Compile: ocamlopt ft_print_comb.ml -o ft_print_comb
 * Run: ./ft_print_comb
 *)

(* unit -> unit *)
let ft_print_comb () =
  let rec loop_third first second third first_printed =
    if third <= 9 then
      (if first < second && second < third then
         (if first_printed then print_string ", " else ();
          print_int first;
          print_int second;
          print_int third;
          loop_third first second (third + 1) true)
       else
         loop_third first second (third + 1) first_printed)
    else
      first_printed
  in
  let rec loop_second first second first_printed =
    if second <= 9 then
      let updated_printed = loop_third first second 0 first_printed in
      loop_second first (second + 1) updated_printed
    else
      first_printed
  in
  let rec loop_first first first_printed =
    if first <= 9 then
      let updated_printed = loop_second first 0 first_printed in
      loop_first (first + 1) updated_printed
    else
      first_printed
  in
  let _ = loop_first 0 false in
  print_char '\n'

let () =
  ft_print_comb ()

